GRDB/Core/Database.swift

```diff
@@ -1243,14 +1243,72 @@ public final class Database: CustomStringConvertible, CustomDebugStringConvertib
     
     // MARK: - Backup
     
-    func backup(
-        to dbDest: Database,
+    /// Copies the database contents into another database.
+    ///
+    /// The `backup` method blocks the current thread until the destination
+    /// database contains the same contents as the source database.
+    ///
+    /// Usage:
+    ///
+    ///     let source: DatabaseQueue = ...
+    ///     let destination: DatabaseQueue = ...
+    ///     try source.write { sourceDb in
+    ///         try destination.barrierWriteWithoutTransaction { destDb in
+    ///             try sourceDb.backup(to: destDb)
+    ///         }
+    ///     }
+    ///
+    ///
+    /// When you're after progress reporting during backup, you'll want to
+    /// perform the backup in several steps. Each step copies the number of
+    /// _database pages_ you specify. See <https://www.sqlite.org/c3ref/backup_finish.html>
+    /// for more information:
+    ///
+    ///     // Backup with progress reporting
+    ///     try sourceDb.backup(
+    ///         to: destDb,
+    ///         pagesPerStep: ...,
+    ///         progress: { (completedPageCount, totalPageCount) in
+    ///            print("\(completedPageCount) pages copied out of \(totalPageCount)")
+    ///         })
+    ///
+    /// The `progress` callback will be called at least once—when
+    /// `completedPageCount == totalPageCount`. If `progress` throws
+    /// when `completedPageCount < totalPageCount`, the backup is aborted
+    /// and the error is rethrown. An error thrown from `progress` when
+    /// `completedPageCount == totalPageCount` is silently ignored.
+    ///
+    /// - parameters:
+    ///     - destDb: The destination database.
+    ///     - pagesPerStep: The number of database pages copied on each backup
+    ///       step. By default, all pages are copied in one single step.
+    ///     - progress: An optional function that is notified of the backup
+    ///       progress.
+    ///     - completedPageCount: The number of copied pages.
+    ///     - totalPageCount: The total number of pages to be copied.
+    /// - throws: The error thrown by `progress` if the backup is abandoned, or
+    ///   any `DatabaseError` that would happen while performing the backup.
+    public func backup(
+        to destDb: Database,
+        pagesPerStep: Int32 = -1,
```

Comment:
I'm thinking we might want to avoid the C-style sentinel value and make this

```swift
pagesPerStep: Int32? = nil
```

instead. When `pagesPerStep == nil`, we pass -1 to `sqlite3_backup_step`. Otherwise we pass `pagesPerStep`. Thoughts? I'm a little wary of making `pagesPerStep == nil` mean "backup in one step", though using `-1` to mean that is not very Swifty either.

Reply:
I understand your question. I would have clearly preferred `nil` a few years ago, but I now prefer to leave raw SQLite concepts alone, unmodified, especially for advanced/expert apis. This helps users who already know SQLite, because they don't have to learn anything new. `pagesPerStep` is a clearly a raw SQLite concept, better understood by reading the SQLite doc. `-1` is a negative value that matches this doc, when `nil` would need an extra interpretation step. And since `-1` is a default value, it won't bother non-experts. To me, `-1` is the winner!

Reply:
WFM